flocker/provision/_digitalocean.py

```diff
@@ -0,0 +1,370 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+DigitalOcean provisioner.
+"""
+import time
+from functools import partial
+
+from ._libcloud import LibcloudProvisioner
+from ._install import (
+    provision, run,
+    task_install_digitalocean_kernel, DIGITALOCEAN_KERNEL
+)
+from ._common import Kernel
+
+
+def retry_on_error(error_checkers, callable, *args, **kwargs):
+    """
+    This function repeats the API call if it raises an exception and if that
+    exception is validated by any of the supplied checkers.
+    It returns the result if the call eventually succeeds.
+
+    :param error_checkers: A ``list`` of ``callables`` which will check for
+        expected exceptions.
+    :param callable: The API function to call.
+    :param args: Positional arguments to supply when calling it.
+    :param kwargs: Keyword arguments to supply when calling it.
+    :return: The result of calling  ``callable``.
+    """
+    while True:
```

Comment:
Can't easily use loop_until because it returns a deferred and all this code is synchronous. But I agree that this is all over complicated. I've simplified things I hope.